asserting throwing for async publish with currently errored batch

#### Environment details

  - OS type and version: linux
  - Python version: `python --version` : Python 3.7.6
  - pip version: `pip --version` :  pip 19.3.1
  - `google-cloud-pubsub` version: `pip show google-cloud-pubsub` 1.3.1

#### Code example

```python
from google.cloud import pubsub_v1
import time


def main():
    publisher = pubsub_v1.PublisherClient()
    topic_path = "projects/*****/topics/*****"

    try:
        publisher.publish(topic_path, data=b'')  # invalid
    except:
        pass

    while True:
        print('publishing')
        publisher.publish(topic_path, data=b'valid')
        time.sleep(1)


if __name__ == '__main__':
    main()
```

#### Stack trace
```
Traceback (most recent call last):
  File "pubsub_error.py", line 21, in <module>
    main()
  File "pubsub_error.py", line 16, in main
    publisher.publish(topic_path, data=b'valid')
  File "/Users/camerondavison/Library/Python/3.7/lib/python/site-packages/google/cloud/pubsub_v1/publisher/client.py", line 364, in publish
    future = sequencer.publish(message)
  File "/Users/camerondavison/Library/Python/3.7/lib/python/site-packages/google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py", line 113, in publish
    future = batch.publish(message)
  File "/Users/camerondavison/Library/Python/3.7/lib/python/site-packages/google/cloud/pubsub_v1/publisher/_batch/thread.py", line 334, in publish
    ), "Publish after stop() or publish error."
AssertionError: Publish after stop() or publish error.
```

As I mentioned just now in https://github.com/googleapis/python-pubsub/pull/26#issuecomment-601333328 it looks like there is an assert that was added that can now trigger for an errored out batch.

batch errors could happen any time so it seems like it would be better to move the assert below the check to see if the batch is still accepting new messages so that a new batch can be created if it errored out.

I validated the same code does not throw an exception on publish in `1.2.0`

--- a/google/cloud/pubsub_v1/publisher/_batch/thread.py
+++ b/google/cloud/pubsub_v1/publisher/_batch/thread.py
@@ -369,3 +369,6 @@ class Batch(base.Batch):
             self.commit()
 
         return future
+
+    def _set_status(self, status):
+        self._status = status

--- a/google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py
+++ b/google/cloud/pubsub_v1/publisher/_sequencer/unordered_sequencer.py
@@ -65,6 +65,12 @@ class UnorderedSequencer(base.Sequencer):
         if self._current_batch:
             self._current_batch.commit()
 
+            # At this point, we lose track of the old batch, but we don't
+            # care since we just committed it.
+            # Setting this to None guarantees the next publish() creates a new
+            # batch.
+            self._current_batch = None
+
     def unpause(self):
         """ Not relevant for this class. """
         raise NotImplementedError

--- a/tests/unit/pubsub_v1/publisher/sequencer/test_unordered_sequencer.py
+++ b/tests/unit/pubsub_v1/publisher/sequencer/test_unordered_sequencer.py
@@ -18,6 +18,7 @@ import pytest
 from google.auth import credentials
 from google.cloud.pubsub_v1 import publisher
 from google.cloud.pubsub_v1 import types
+from google.cloud.pubsub_v1.publisher._batch import base
 from google.cloud.pubsub_v1.publisher._sequencer import unordered_sequencer
 
 
@@ -101,4 +102,24 @@ def test_publish_batch_full():
     # Will create a new batch since the old one is full, and return a future.
     future = sequencer.publish(message)
     batch.publish.assert_called_once_with(message)
+    assert future is not None
+
+
+def test_publish_after_batch_error():
+    client = create_client()
+    message = create_message()
+    batch = mock.Mock(spec=client._batch_class)
+
+    sequencer = unordered_sequencer.UnorderedSequencer(client, "topic_name")
+    sequencer._set_batch(batch)
+
+    sequencer.commit()
+    batch.commit.assert_called_once()
+
+    # Simulate publish RPC failing.
+    batch._set_status(base.BatchStatus.ERROR)
+
+    # Will create a new batch since the old one has been committed. The fact
+    # that it errored should not matter in the publish of the next message.
+    future = sequencer.publish(message)
     assert future is not None

--- a/tests/unit/pubsub_v1/publisher/sequencer/test_unordered_sequencer.py
+++ b/tests/unit/pubsub_v1/publisher/sequencer/test_unordered_sequencer.py
@@ -120,6 +120,7 @@ def test_publish_after_batch_error():
     batch._set_status(base.BatchStatus.ERROR)
 
     # Will create a new batch since the old one has been committed. The fact
-    # that it errored should not matter in the publish of the next message.
+    # that the old batch errored should not matter in the publish of the next
+    # message.
     future = sequencer.publish(message)
     assert future is not None

